langchain/vectorstores/deeplake.py

```diff
@@ -3,40 +3,72 @@
 
 import logging
 import uuid
-from typing import Any, Iterable, List, Optional, Sequence
+from typing import Any, Iterable, List, Tuple, Optional, Sequence, Dict
 
 import numpy as np
+from functools import partial
 
 from langchain.docstore.document import Document
 from langchain.embeddings.base import Embeddings
 from langchain.vectorstores.base import VectorStore
+from langchain.vectorstores.utils import maximal_marginal_relevance
 
 logger = logging.getLogger()
 
+distance_metric_map = {
+    'l2': lambda a, b: np.linalg.norm(a - b, axis=1, ord=2),
+    'l1': lambda a, b: np.linalg.norm(a - b, axis=1, ord=1),
+    'max': lambda a, b: np.linalg.norm(a - b, axis=1, ord=np.inf),
+    'cos': lambda a, b: np.dot(a, b.T) / (np.linalg.norm(a) * np.linalg.norm(b, axis=1)),
+    'dot': lambda a, b: np.dot(a, b.T),
+}
+
+
+def search(
+    query_embedding: np.ndarray,
+    data_vectors: np.ndarray,
+    k: int = 4,
+    distance_metric: str = 'L2',
+) -> Tuple[List, List]:
+    """Naive search for nearest neighbors
+
+    args: 
+        query_embedding: np.ndarray
+        data_vectors: np.ndarray
+        k (int): number of nearest neighbors
+        distance_metric: distance function 'L2' for Euclidean, 'L1' for Nuclear, 'Max' l-infinity distnace, 'cos' for cosine similarity, 'dot' for dot product
+
+    returns:
+        nearest_indices: List, indices of nearest neighbors
+    """
+    # Calculate the distance between the query_vector and all data_vectors
+    distances = distance_metric_map[distance_metric](query_embedding, data_vectors)
+    nearest_indices = np.argsort(distances)
+    nearest_indices = nearest_indices[::-1][:k]\
+        if distance_metric in ['cos'] else nearest_indices[:k]
+
+    return nearest_indices.tolist(), distances[nearest_indices].tolist()
 
-def L2_search(
-    query_embedding: np.ndarray, data_vectors: np.ndarray, k: int = 4
-) -> list:
-    """naive L2 search for nearest neighbors"""
-    # Calculate the L2 distance between the query_vector and all data_vectors
-    distances = np.linalg.norm(data_vectors - query_embedding, axis=1)
 
-    # Sort the distances and return the indices of the k nearest vectors
-    nearest_indices = np.argsort(distances)[:k]
-    return nearest_indices.tolist()
+def dp_filter(x: dict, filter: Optional[Dict[str, str]] = None) -> bool:
+    """Filter helper function for Deep Lake"""
+    metadata = x['metadata'].data()["value"]
+    if any([str(v) != str(metadata[k]) if k in metadata else True for k, v in filter.items()]):
+        return False
+    return True
```

Comment:
Refac:

(Also, why str conversion?)
```suggestion
def dp_filter(x: Dict, filter: Optional[Dict[str, str]] = None) -> bool:
    """Filter helper function for Deep Lake"""
    metadata = x['metadata'].data()["value"]
    return all(k in metadata and str(v) == str(metadata[k]) for k, v in filter.items())
```

Reply:
Good point!